File: src/main/java/io/xol/chunkstories/api/voxel/components/VoxelComponent.java
package io.xol.chunkstories.api.voxel.components;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

import io.xol.chunkstories.api.net.packets.PacketVoxelUpdate;
import io.xol.chunkstories.api.server.RemotePlayer;
import io.xol.chunkstories.api.world.cell.CellComponents;
import io.xol.chunkstories.api.world.chunk.WorldUser;
import io.xol.chunkstories.api.world.serialization.StreamSource;
import io.xol.chunkstories.api.world.serialization.StreamTarget;

public abstract class VoxelComponent {
	private final CellComponents holder;
	
	public VoxelComponent(CellComponents holder) {
		this.holder = holder;
	}
	
	public final CellComponents holder() {
		return holder;
	}
	
	public final String name() {
		return holder.name(this);
	}
	
	/** Pushes the component to every client subscribed to the chunk owning this voxel */
	public void pushComponentEveryone() {
		for(WorldUser user : holder.users()) {
			if(user instanceof RemotePlayer) {
				pushComponent((RemotePlayer)user);
			}
		}
	}
	
	/** Pushes the component to a specific player */
	public void pushComponent(RemotePlayer player) {
		PacketVoxelUpdate packet = new PacketVoxelUpdate(holder.cell(), this);
		player.pushPacket(packet);
	}
	
	public abstract void push(StreamTarget destinator, DataOutputStream dos) throws IOException;

	public abstract void pull(StreamSource from, DataInputStream dis) throws IOException;
}

File: src/main/java/xyz/chunkstories/api/voxel/components/VoxelComponent.kt
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.api.voxel.components

import xyz.chunkstories.api.content.json.Json
import java.io.IOException

import xyz.chunkstories.api.net.packets.PacketVoxelUpdate
import xyz.chunkstories.api.server.RemotePlayer

/**
 * Holds extra information on top of the 32 bits every getCell gets.
 * Has to provide it's own serialization routines
 */
abstract class VoxelComponent(val holder: CellComponents) {

    val name: String
        get() = holder.getRegisteredComponentName(this) ?: throw Exception("Assertion broken")

    /** Pushes the component to every client subscribed to the chunk owning this voxel  */
    fun pushComponentEveryone() {
        for (user in holder.chunk.holder.users) {
            if (user is RemotePlayer) {
                pushComponent(user)
            }
        }
    }

    /** Pushes the component to a specific player  */
    fun pushComponent(player: RemotePlayer) {
        val packet = PacketVoxelUpdate(holder.cell, this)
        player.pushPacket(packet)
    }

    @Throws(IOException::class)
    abstract fun serialize() : Json?

    @Throws(IOException::class)
    abstract fun deserialize(json: Json)
}